### PR TITLE
Introduce MutableOption

### PR DESCRIPTION
**Pull request by charleso**
@markhibberd This OK?

### CODE REVIEW

File: ivory-mr/src/main/scala/com/ambiata/ivory/mr/MutableOption.scala [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ *   u$$$uuu$$$$$$$$$uu ""$$$$$$$$$$$uuu$$$
+ *   $$$$$$$$$$""""           ""$$$$$$$$$$$"
+ *    "$$$$$"                      ""$$$$""
+ *      $$$"                         $$$$"
+ * </pre>
+ *
+ * This is to represent an optional value in cases where we wish to avoid the cost of an extra allocation.
+ * Please use with extreme care!
+ */
+class MutableOption[A](private var value: A, var isSet: Boolean) {
```
Comment (charleso) [inline ivory-mr/src/main/scala/com/ambiata/ivory/mr/MutableOption.scala]: (Or is `OptionMutable` better?)

File: ivory-mr/src/main/scala/com/ambiata/ivory/mr/MutableOption.scala [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  def get: A = {
+    if (!isSet)
+      Crash.error(Crash.DataIntegrity, "The value is currently unset, please check `set` first")
+    value
+  }
+}
+
+object MutableOption {
+
+  def none[A](v: A): MutableOption[A] =
```
Comment (charleso) [inline ivory-mr/src/main/scala/com/ambiata/ivory/mr/MutableOption.scala]: Is there a way to avoid this somehow?

Comment (markhibberd) [inline ivory-mr/src/main/scala/com/ambiata/ivory/mr/MutableOption.scala]: You can covariant hack it. But I don't think it is necessary, but you probably want a `setNone` method so you can tweak an existing value to be none.

Comment (charleso) [inline ivory-mr/src/main/scala/com/ambiata/ivory/mr/MutableOption.scala]: > but you probably want a setNone method so you can tweak an existing value to be none.

Whoops. Yes I do.

File: ivory-mr/src/main/scala/com/ambiata/ivory/mr/MutableOption.scala (hunk 1 above)
Comment (markhibberd) [inline ivory-mr/src/main/scala/com/ambiata/ivory/mr/MutableOption.scala]: Personally not adverse to just making it unicode skull and cross bones. But this is fine.

Comment (markhibberd) [inline ivory-mr/src/main/scala/com/ambiata/ivory/mr/MutableOption.scala]: `case class` or `toString`/`equals` implementations. Prefer case class though, I don't think there is a good reason not to IMO.

File: ivory-mr/src/main/scala/com/ambiata/ivory/mr/MutableOption.scala [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+ * Please use with extreme care!
+ */
+class MutableOption[A](private var value: A, var isSet: Boolean) {
+
+  def get: A = {
+    if (!isSet)
+      Crash.error(Crash.DataIntegrity, "The value is currently unset, please check `set` first")
+    value
+  }
+}
```
Comment (markhibberd) [inline ivory-mr/src/main/scala/com/ambiata/ivory/mr/MutableOption.scala]: A `toOption` method would be a nice addition, probably end up useful for testing.

**Comment by markhibberd** [comment]
:+1:

File: ivory-mr/src/main/scala/com/ambiata/ivory/mr/MutableOption.scala (hunk 3 above)
Comment (charleso) [inline ivory-mr/src/main/scala/com/ambiata/ivory/mr/MutableOption.scala]: > A toOption method would be a nice addition, probably end up useful for testing.

Ha. I had that and then removed it. Happy to put it back.

**Pull request merged**